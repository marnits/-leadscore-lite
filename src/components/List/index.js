import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import Header from '../Header';
import Label from '../Label';
import Text from '../Text';
import Spinner from '../Spinner';
import styles from './styles.module.scss';
import columnType from '../../utils/propTypes/column';

const isScrolledToBottom = el => el.scrollHeight - el.scrollTop === el.clientHeight;
const isCloseToTheBottom = el => el.scrollHeight - el.scrollTop <= el.clientHeight + 100;

class List extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      bottomReached: false,
      showShadow: true,
    };

    this.assignScrollableRef = this.assignScrollableRef.bind(this);
  }

  static getDerivedStateFromProps({ last }, { bottomReached }) {
    return last
      ? {
        showShadow: !bottomReached,
      }
      : null;
  }

  shouldLoadMoreItems = () => {
    const { last, loading } = this.props;

    return isCloseToTheBottom(this.scrollable) && !last && !loading;
  };

  calculateShadowWidth = () => (
    this.scrollable
      ? `${this.scrollable.clientWidth}px`
      : 0
  );

  handleScroll = () => {
    const { bottomReached } = this.state;
    const { loadMore } = this.props;

    if (this.shouldLoadMoreItems()) {
      loadMore();
    }

    if (isScrolledToBottom(this.scrollable) !== bottomReached) {
      this.setState({
        bottomReached: isScrolledToBottom(this.scrollable),
      });
    }
  };

  passPropsToChildren = (children, columns) => (
    React.Children.map(children, child => (
      React.cloneElement(child, { columns })
    ))
  );

  renderEmptyElements = () => {
    const { loading } = this.props;
    return !loading && (
      <div className={styles.empty}>
        <Text type="subdued" text="There are no elements" />
      </div>
    );
  }

  assignScrollableRef(scrollable) {
    this.scrollable = scrollable;
  }

  render() {
    const {
      loading,
      title,
      columns,
      children,
    } = this.props;
    const { showShadow } = this.state;

    return (
      <div className={styles.wrapper}>
        {title && (
          <div className={styles.title}>
            <Header text={title} />
          </div>
        )}
        <div className={styles.row}>
          {columns.map(column => (
            <div
              className={`
                ${styles['column-name']}
                ${styles[`align-content-${column.align}`]}
                ${styles[`flex-${column.size}`]}
              `}
              key={column.name}
            >
              <Label type="small" text={column.name} />
            </div>
          ))}
        </div>
        <div
          ref={this.assignScrollableRef}
          className={styles.list}
          onScroll={this.handleScroll}
        >
          {children && children.length !== 0
            ? this.passPropsToChildren(children, columns)
            : this.renderEmptyElements()
          }
          {loading && <Spinner />}
        </div>
        {(showShadow && !loading) && (
          <div
            className={styles.shadow}
            style={{ width: this.calculateShadowWidth() }}
          />
        )}
      </div>
    );
  }
}

List.propTypes = {
  last: PropTypes.bool.isRequired,
  loading: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  columns: PropTypes.arrayOf(columnType).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
  loadMore: PropTypes.func.isRequired,
};

List.defaultProps = {
  children: null,
};

export default List;
