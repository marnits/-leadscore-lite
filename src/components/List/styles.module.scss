.wrapper {
  position: relative;
  display: flex;
  flex-direction: column;
  width: 100%;
  border-bottom: 1px solid #dedede;

  .title {
    padding-bottom: 16px;
    border-bottom: 1px solid #dedede;
  }

  .list {
    background-color: #fbfcfd;
    flex-grow: 1;
    overflow-y: auto;
  }

  .column-name {
    padding: 8px 16px;
  }

  .row {
    display: flex;
    flex-direction: row;
    min-height: 38px;
  }

  .empty {
    border: 1px solid #dedede;
    border-bottom: 0;
    border-radius: 1px;
    padding: 16px;
    box-shadow: inset 0 0 0 0 #dedede;
    outline: none;
    text-align: center;
  }

  .shadow {
    background-image: linear-gradient(to bottom, rgba(255, 255, 255, 0), #ffffff);
    width: 100%;
    position: absolute;
    height: 71px;
    bottom: 0;
    pointer-events: none;
  }

  .align-content- {
    &left {
      text-align: left;
    }

    &center {
      text-align: center;
    }

    &right {
      text-align: right;
    }
  }

  .flex-1 { flex: 1 }
  .flex-2 { flex: 2 }
  .flex-3 { flex: 3 }
}
