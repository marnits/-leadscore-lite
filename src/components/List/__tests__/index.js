import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import List from '../index';

const props = {
  title: 'Test list',
  loading: false,
  last: false,
  columns: [
    {
      name: 'test1',
    },
  ],
  loadMore: () => {},
};

describe('List', () => {
  test('should call assignScrollableRef method', () => {
    const spy = jest.spyOn(List.prototype, 'assignScrollableRef')
      .mockImplementation(function mock() {
        this.scrollable = {
          offsetHeight: 100,
          scrollTop: 100,
          clientHeight: 100,
        };
      });

    mount(<List {...props} />);

    expect(spy).toHaveBeenCalled();
  });

  test('should not return 0 while calling calculateShadowWidth method when not mounted', () => {
    const wrapper = shallow(<List {...props} />);

    const calculate = wrapper.instance().calculateShadowWidth();

    expect(calculate).toBe(0);
  });

  test('should not return 0 while calling calculateShadowWidth method when not mounted', () => {
    const wrapper = shallow(<List {...props} />);

    const calculate = wrapper.instance().calculateShadowWidth();

    expect(calculate).toBe(0);
  });

  test('should test calculateShadowWidth method', () => {
    jest.spyOn(List.prototype, 'assignScrollableRef')
      .mockImplementation(function mock() {
        this.scrollable = {
          offsetHeight: 300,
          scrollTop: 200,
          clientHeight: 100,
          clientWidth: 1,
        };
      });

    const wrapper = mount(<List {...props} />);

    const calculate = wrapper.instance().calculateShadowWidth();

    expect(calculate).toBe('1px');
  });

  test('should not load items when scrolled not enough', () => {
    const spy = jest.spyOn(List.prototype, 'assignScrollableRef')
      .mockImplementation(function mock() {
        this.scrollable = {
          offsetHeight: 300,
          scrollTop: 200,
          clientHeight: 100,
        };
      });

    const wrapper = mount(<List {...props} />);

    const shouldLoadMore = wrapper.instance().shouldLoadMoreItems();

    expect(spy).toHaveBeenCalled();
    expect(shouldLoadMore).toBe(false);
  });

  test('should not load items when still loading', () => {
    const spy = jest.spyOn(List.prototype, 'assignScrollableRef')
      .mockImplementation(function mock() {
        this.scrollable = {
          offsetHeight: 0,
          scrollTop: 0,
          clientHeight: 0,
        };
      });

    const wrapper = mount(<List {...props} loading />);

    const shouldLoadMore = wrapper.instance().shouldLoadMoreItems();

    expect(spy).toHaveBeenCalled();
    expect(shouldLoadMore).toBe(false);
  });

  test('should load items when scrolled enough', () => {
    const spy = jest.spyOn(List.prototype, 'assignScrollableRef')
      .mockImplementation(function mock() {
        this.scrollable = {
          scrollHeight: 0,
          scrollTop: 0,
          clientHeight: 100,
        };
      });

    const wrapper = mount(<List {...props} />);

    const shouldLoadMore = wrapper.instance().shouldLoadMoreItems();

    expect(spy).toHaveBeenCalled();
    expect(shouldLoadMore).toBe(true);
  });

  test('should scolled and not set or call anything', () => {
    const load = jest.spyOn(props, 'loadMore');
    jest.spyOn(List.prototype, 'assignScrollableRef')
      .mockImplementation(function mock() {
        this.scrollable = {
          offsetHeight: 300,
          scrollTop: 200,
          clientHeight: 100,
        };
      });

    const wrapper = mount(<List {...props} />);

    expect(wrapper.state('bottomReached')).toBe(false);

    wrapper.instance().handleScroll();

    expect(wrapper.state('bottomReached')).toBe(false);
    expect(load).not.toHaveBeenCalled();
  });

  test('should call loadMore when scolled close to the bottom', () => {
    const load = jest.spyOn(props, 'loadMore');
    const spy = jest.spyOn(List.prototype, 'assignScrollableRef')
      .mockImplementation(function mock() {
        this.scrollable = {
          scrollHeight: 0,
          scrollTop: 0,
          clientHeight: 100,
        };
      });

    const wrapper = mount(<List {...props} />);

    wrapper.instance().handleScroll();

    expect(spy).toHaveBeenCalled();
    expect(load).toHaveBeenCalled();
  });

  test('should set state properly when bottom reached', () => {
    jest.spyOn(List.prototype, 'assignScrollableRef')
      .mockImplementation(function mock() {
        this.scrollable = {
          scrollHeight: 300,
          scrollTop: 200,
          clientHeight: 100,
        };
      });

    const wrapper = mount(<List {...props} />);

    expect(wrapper.state('bottomReached')).toBe(false);

    wrapper.instance().handleScroll();

    expect(wrapper.state('bottomReached')).toBe(true);
  });

  test('should test static method called getDerivedStateFromProps', () => {
    expect(List.getDerivedStateFromProps({ last: false }, { bottomReached: true }))
      .toBe(null);
    expect(List.getDerivedStateFromProps({ last: true }, { bottomReached: true }))
      .toEqual({ showShadow: false });
  });

  test('should properly render list without children', () => {
    const tree = renderer.create(<List {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
