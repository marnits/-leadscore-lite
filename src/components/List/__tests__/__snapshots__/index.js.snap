// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should properly render list without children 1`] = `
<div
  className="wrapper"
>
  <div
    className="title"
  >
    <span
      className="header"
    >
      Test list
    </span>
  </div>
  <div
    className="row"
  >
    <div
      className="
        column-name
        align-content-left
        flex-1
      "
    >
      <label
        className="label"
        htmlFor={null}
      >
        test1
      </label>
    </div>
  </div>
  <div
    className="list"
    onScroll={[Function]}
  >
    <div
      className="empty"
    >
      <span
        className="text false"
      >
        There are no elements
      </span>
    </div>
  </div>
  <div
    className="shadow"
    style={
      Object {
        "width": 0,
      }
    }
  />
</div>
`;
